### PR TITLE
C API documentation

### PR DESCRIPTION
**Pull request by dalg24**
Will be completed when Doxygen documentation in [DTK_C_API.h](packages/Interface/src/DTK_C_API.h) has been updated.

This documentation is published [here](http://datatransferkitdalg24.readthedocs.io/en/latest/api.html#)

To generate the documentation on your machine run this from DTK root directory:
```bash
$ docker run --rm -v $PWD:/scratch -w /scratch dalg24/clang-format bash -c "sphinx-build docs/source docs/build"
```
Then open `docs/build/index.html`.

Connect to #252

### CODE REVIEW

**Comment by aprokop** [comment]
Can you do the same with Fortran?

**Comment by dalg24** [comment]
> Can you do the same with Fortran?

Yes, Oxygen is compatible with Fortran.  I just tried rebasing onto your branch with the Fortran API  (I added `".f90"` in the list of extensions to generate Doxygen XML output, and imported the source file into our Sphinx documentation) and it worked just fine.  See below:
```diff
diff --git a/docs/source/api.rst b/docs/source/api.rst
index 0df10dd9..11a91455 100644
--- a/docs/source/api.rst
+++ b/docs/source/api.rst
@@ -28,4 +28,4 @@ C API
 Fortran API
 -----------

-Under construction.
+.. doxygenfile:: DTK_Fortran_API.f90
diff --git a/docs/source/conf.py b/docs/source/conf.py
index 264cf67f..a497e69f 100644
--- a/docs/source/conf.py
+++ b/docs/source/conf.py
@@ -26,7 +26,7 @@ else:
     source_directories = [
         "packages/Interface/src",
     ]
-    file_extensions = [".h", ".hpp", ".cpp"]
+    file_extensions = [".f90", ".h", ".hpp", ".cpp"]
     # glob files with extensions for given paths
     source_files = []
     for _dir in source_directories:
```

**Comment by dalg24** [comment]
Note that this would expose the SWIG internals so we would probably want to use `DOXYGEN_SHOULD_SKIP_THIS` to ignore some of the code.  Also not sure how well it is supported by Breathe (the Shinx extension plugin we use to import the Doxygen doc).

**Comment by dalg24** [comment]
I added documentation for the initialize/finalize functions.  This current version is rebased on #253 
DO NOT MERGE AS IS.

File: packages/Interface/src/DTK_C_API.h [hunk 1]
```diff
@@ -28,195 +28,513 @@
...
+ *
+ *  Register with a user application using DTK_set_function() by passing
+ *  DTK_MIXED_TOPOLOGY_CELL_LIST_SIZE_FUNCTION as the \p type argument.
+ *
+ *  \param[in] user_data Pointer to custom user data.
+ *  \param[out] space_dim Spatial dimension.
+ *  \param[out] local_num_nodes Number of nodes DTK will allocate memory for.
+ *  \param[out] local_num_cells Number of cells DTK will allocate memory for.
+ *  \param[out] total_nodes_per_cell TODO
+ *  \param[out] has_gost Whether some of the cells that will be passed are
```
Comment (Rombur) [inline packages/Interface/src/DTK_C_API.h]: typo has_gost

File: packages/Interface/src/DTK_C_API.h [hunk 2]
```diff
@@ -28,195 +28,513 @@
...
+/** \brief Prototype function to get the size parameters for building a node
+ *         list.
+ *
+ *  Register with a user application using DTK_set_function() by passing
+ *  DTK_NODE_LIST_SIZE_FUNCTION as the \p type argument.
+ *
+ *  \param[in] user_data Pointer to custom user data.
+ *  \param[out] space_dim Spatial dimension.
+ *  \param[out] local_num_nodes Number of nodes DTK will allocate memory for.
+ *  \param[out] has_gost Whether some of the nodes that will be passed are
```
Comment (Rombur) [inline packages/Interface/src/DTK_C_API.h]: typo has_gost ther is the typo several time

File: packages/Interface/src/DTK_C_API.h [hunk 3]
```diff
@@ -28,195 +28,513 @@
...
                                                       bool *has_ghosts );
 
-/*! Get the data for a bounding volume list. */
+/** \brief Prototype function to get the data for a bounding volume list.
+ *
+ *  Register with a user application using DTK_set_function() by passing
+ *  DTK_BOUNDING_VOLUME_LIST_DATA_FUNCTION as the \p type argument.
+ *
+ *  \param[in] user_data Pointer to custom user data.
+ *  \param[out] bounding_volumes Bounding volumes..
```
Comment (Rombur) [inline packages/Interface/src/DTK_C_API.h]: extra dot

File: packages/Interface/src/DTK_C_API.h [hunk 4]
```diff
@@ -28,195 +28,513 @@
...
+ *  polyhedron list.
+ *
+ *  Register with a user application using DTK_set_function() by passing
+ *  DTK_POLYHEDRON_LIST_SIZE_FUNCTION as the \p type argument.
+ *
+ *  \param[in] user_data Pointer to custom user data.
+ *  \param[out] space_dim Spatial dimension.
+ *  \param[out] local_num_nodes Number of nodes DTK will allocate memory for.
+ *  \param[out] local_num_faces Number of faces DTK will allocate memory for.
+ *  \param[out] total_nodes_per_face TODO
```
Comment (Rombur) [inline packages/Interface/src/DTK_C_API.h]: Don't forget about these

File: packages/Interface/src/DTK_C_API.h [hunk 5]
```diff
@@ -28,195 +28,513 @@
...
- * per object.
+/** \brief Prototype function to get the size data for a degree-of-freedom id
+ *  map with a single number of dofs per object.
+ *
+ *  Register with a user application using DTK_set_function() by passing
+ *  DTK_DOF_MAP_DATA_FUNCTION as the \p type argument.
+ *
+ *  \param[in] user_data Pointer to custom user data.
+ *  \param[out] global_dof_ids Globally unique ids for DOFs on this process.
+ *  \param[out] object_dof_ids Local object IDs.
```
Comment (Rombur) [inline packages/Interface/src/DTK_C_API.h]: Which object are we talking about?

**Comment by dalg24** [comment]
OK, I rebased on master and cleaned things up.  This is ready for review/merge.

You can view the doc [here](http://datatransferkitdalg24.readthedocs.io/en/latest/api.html#)

File: packages/Interface/src/DTK_C_API.h [hunk 6]
```diff
@@ -28,195 +28,525 @@
...
+    DTK_PULL_FIELD_DATA_FUNCTION /** See DTK_PullFieldDataFunction() */,
+    DTK_PUSH_FIELD_DATA_FUNCTION /** See DTK_PushFieldDataFunction() */,
+    DTK_EVALUATE_FIELD_FUNCTION /** See DTK_EvaluateFieldFunction() */,
 } DTK_FunctionType;
+// clang-format on /////////////////////////////////////////////////////////////
 
-/*! Register a function with DTK. */
+/** \brief Register a function as a callback.
+ *
+ *  This registers a custom function as a callback for DTK to communcate with
```
Comment (Rombur) [inline packages/Interface/src/DTK_C_API.h]: typo communicate

File: packages/Interface/src/DTK_C_API.h [hunk 7]
```diff
@@ -28,195 +28,525 @@
...
+ *  number of dofs (e.g. mixed topology cell lists or polyhedron lists).
+ *
+ *  Register with a user application using DTK_set_function() by passing
+ *  DTK_MIXED_TOPOLOGY_DOF_MAP_SIZE_FUNCTION as the \p type argument.
+ *
+ *  \param[in] user_data Pointer to custom user data.
+ *  \param[out] local_num_dofs Number of degrees of freedom owned by this
+ *  process.
+ *  \param[out] local_num_objects Number of objects on this process.
+ *  \param[out] total_dofs_per_objects Total degrees of freedom per pbjects.
```
Comment (Rombur) [inline packages/Interface/src/DTK_C_API.h]: typo objects

**Pull request merged**